data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriters.java

```diff
@@ -0,0 +1,418 @@
+    }
+  }
+
+  private static class DateWriter implements OrcValueWriter<LocalDate> {
+    private static final OrcValueWriter<LocalDate> INSTANCE = new DateWriter();
+
+    @Override
+    public Class<LocalDate> getJavaClass() {
+      return LocalDate.class;
+    }
+
+    @Override
+    public void nonNullWrite(int rowId, LocalDate data, ColumnVector output) {
+      ((LongColumnVector) output).vector[rowId] = ChronoUnit.DAYS.between(EPOCH_DAY, data);
+    }
+  }
+
+  private static class TimestampTzWriter implements OrcValueWriter<OffsetDateTime> {
+    private static final OrcValueWriter<OffsetDateTime> INSTANCE = new TimestampTzWriter();
+
+    @Override
+    public Class<OffsetDateTime> getJavaClass() {
+      return OffsetDateTime.class;
+    }
+
+    @Override
+    public void nonNullWrite(int rowId, OffsetDateTime data, ColumnVector output) {
+      TimestampColumnVector cv = (TimestampColumnVector) output;
+      cv.time[rowId] = data.toInstant().toEpochMilli(); // millis
+      cv.nanos[rowId] = (data.getNano() / 1_000) * 1_000; // truncate nanos to only keep microsecond precision
+    }
+  }
+
+  private static class TimestampWriter implements OrcValueWriter<LocalDateTime> {
+    private static final OrcValueWriter<LocalDateTime> INSTANCE = new TimestampWriter();
+
+    @Override
+    public Class<LocalDateTime> getJavaClass() {
+      return LocalDateTime.class;
+    }
+
+    @Override
+    public void nonNullWrite(int rowId, LocalDateTime data, ColumnVector output) {
+      TimestampColumnVector cv = (TimestampColumnVector) output;
+      cv.setIsUTC(true);
+      cv.time[rowId] = data.toInstant(ZoneOffset.UTC).toEpochMilli(); // millis
+      cv.nanos[rowId] = (data.getNano() / 1_000) * 1_000; // truncate nanos to only keep microsecond precision
+    }
+  }
+
+  private static class Decimal18Writer implements OrcValueWriter<BigDecimal> {
+    private final int scale;
+
+    Decimal18Writer(int scale) {
+      this.scale = scale;
+    }
+
+    @Override
+    public Class<BigDecimal> getJavaClass() {
+      return BigDecimal.class;
+    }
+
+    @Override
+    public void nonNullWrite(int rowId, BigDecimal data, ColumnVector output) {
+      // TODO: validate precision and scale from schema
+      ((DecimalColumnVector) output).vector[rowId]
+          .setFromLongAndScale(data.unscaledValue().longValueExact(), scale);
+    }
+  }
+
+  private static class Decimal38Writer implements OrcValueWriter<BigDecimal> {
+    private static final OrcValueWriter<BigDecimal> INSTANCE = new Decimal38Writer();
+
+    @Override
+    public Class<BigDecimal> getJavaClass() {
+      return BigDecimal.class;
+    }
+
+    @Override
+    public void nonNullWrite(int rowId, BigDecimal data, ColumnVector output) {
+      // TODO: validate precision and scale from schema
+      ((DecimalColumnVector) output).vector[rowId].set(HiveDecimal.create(data, false));
+    }
+  }
+
+  private static class ListWriter implements OrcValueWriter<List> {
```

Comment:
Missing parameter type.

Reply:
I did not provide the explicit parameter type here, because  [getJavaClass()](https://github.com/apache/iceberg/pull/1197/files#diff-b1b07b15f036000a3f2bed76fdd9f961R357)  will need to return a  `List` class with generic type, while Java don't support this now.   Pls see [here](http://www.angelikalanger.com/GenericsFAQ/FAQSections/ParameterizedTypes.html#FAQ106)

Reply:
We still need to add the parameterized types.

The FAQ entry you pointed to explains why there is no class literal, like `List<String>.class`. All variants of `List` use `List.class` because there is only one concrete type at runtime. But we still want to use type parameters to be explicit about what is passed around.